lib/pure/seqmath.nim

```diff
@@ -0,0 +1,1136 @@
+
+proc `>=`*[T:Point](a, b: T): bool =
+  result = (a.`>`(b) or a.`==`(b))
+
+proc swap*(a: var Point) {.inline.} =
+  ## swap ``x`` and ``y`` inplace
+  let t = a.x
+  a.x = a.y
+  a.y = t
+
+proc swap*(a: Point): Point {.inline.} =
+  ## returns Point ``a`` with ``x`` and ``y`` swapped
+  let t = a.x
+  result.x = a.y
+  result.y = t
+
+
+#proc toFloat*(x: openArray[int]): seq[float] =
+#  ## return a copy of ``x`` with all elements
+#  ## converted to ``float``
+#  result = newSeq[float](x.len)
+#  for i in 0..<x.len:
+#    result[i] = x[i].float
+
+# ----------- nested seq math ---------------------------
+
+template liftScalarProc*(fname) =
+  ## Lift a proc taking one scalar parameter and returning a
+  ## scalar value (eg ``proc sssss[T](x: T): float``),
+  ## to provide templated procs that can handle a single
+  ## parameter of seq[T] or nested seq[seq[]] or the same type
+  ##
+  ## .. code-block:: Nim
+  ##  liftScalarProc(abs)
+  ##  # now abs(@[@[1,-2], @[-2,-3]]) == @[@[1,2], @[2,3]]
+  proc fname[T](x: openarray[T]): auto =
+    var temp: T
+    type outType = type(fname(temp))
+    result = newSeq[outType](x.len)
+    for i in 0..<x.len:
+      result[i] = fname(x[i])
+
+## Available math functions for nested sequences
+##----
+## **From the system module**
+##
+## - toFloat
+## - toInt
+## - toBiggestFloat
+## - toBiggestInt
+## - abs
+##
+## **From the math module**
+##
+## - classify
+## - binom
+## - fac
+## - isPowerOfTwo
+## - nextPowerOfTwo
+## - countBits32
+## - random
+## - sqrt
+## - cbrt
+## - log10
+## - log2
+## - exp
+## - arccos
+## - arcsin
+## - arctan
+## - cos
+## - cosh
+## - sin
+## - sinh
+## - tan
+## - tanh
+## - erf
+## - erfc
+## - lgamma
+## - tgamma
+## - trunc
+## - round
+## - floor
+## - ceil
+## - degToRad
+## - radToDeg
+
+# ---- from system.nim -------------
+liftScalarProc(toFloat)
+liftScalarProc(toInt)
+liftScalarProc(toBiggestFloat)
+liftScalarProc(toBiggestInt)
+liftScalarProc(abs)
+# ---- from math.nim --------------
+liftScalarProc(classify)
+liftScalarProc(binom)
+liftScalarProc(fac)
+liftScalarProc(isPowerOfTwo)
+liftScalarProc(nextPowerOfTwo)
+liftScalarProc(countBits32)
+liftScalarProc(random)
+liftScalarProc(sqrt)
+liftScalarProc(cbrt)
+liftScalarProc(log10)
+liftScalarProc(log2)
+liftScalarProc(exp)
+#liftScalarProc2(fexp)
+liftScalarProc(arccos)
+liftScalarProc(arcsin)
+liftScalarProc(arctan)
+#liftScalarProc2(arctan2)
+liftScalarProc(cos)
+liftScalarProc(cosh)
+#liftScalarProc2(hypot)
+liftScalarProc(sin)
+liftScalarProc(sinh)
+liftScalarProc(tan)
+liftScalarProc(tanh)
+#liftScalarProc2(pow)
+liftScalarProc(erf)
+liftScalarProc(erfc)
+liftScalarProc(lgamma)
+liftScalarProc(tgamma)
+liftScalarProc(trunc)
+liftScalarProc(round)
+liftScalarProc(floor)
+liftScalarProc(ceil)
+liftScalarProc(degToRad)
+liftScalarProc(radToDeg)
+liftScalarProc(gcd)
+liftScalarProc(lcm)
+
+# ----------- cumulative seq math -----------------------
+
+proc cumProd*[T:int|float](x: openArray[T]): seq[T] =
```

Comment:
The T:int/float everywhere seems dangerous. I bet many algorithms work with e.g. bignums just fine. ;-)